cpp/src/io/statistics/column_stats.cu

```diff
@@ -0,0 +1,755 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "column_stats.h"
+#include <math_constants.h>
+#include <io/utilities/block_utils.cuh>
+
+namespace cudf {
+namespace io {
+
+struct stats_state_s
+{
+    stats_column_desc col;
+    statistics_group group;
+    statistics_chunk ck;
+    volatile statistics_val warp_min[32];
+    volatile statistics_val warp_max[32];
+    volatile statistics_val warp_sum[32];
+};
+
+struct merge_state_s
+{
+    stats_column_desc col;
+    statistics_merge_group group;
+    statistics_chunk ck;
+    volatile statistics_val warp_min[32];
+    volatile statistics_val warp_max[32];
+    volatile statistics_val warp_sum[32];
+    volatile uint32_t warp_non_nulls[32];
+    volatile uint32_t warp_nulls[32];
+};
+
+
+inline __device__ int64_t WarpReduceMinInt(int64_t vmin)
```

Comment:
Yeah, but it requires additional scratch state to abstract between pre-SHFL and post-SHFL GPUs. Since we also need more specialized versions for double and strings, I find it more consistent to keep it here and not introduce a dependency on cub.